[mlir][TilingInterface] Use `LoopLikeOpInterface` in tiling using SCF to unify tiling with `scf.for` and `scf.forall`.

MaheshRavishankar opened the pull request:
Using `LoopLikeOpInterface` as the basis for the implementation unifies all the tiling logic for both `scf.for` and `scf.forall`. The only difference is the actual loop generation. This is a follow up to https://github.com/llvm/llvm-project/pull/72178
Instead of many entry points for each loop type, the loop type is now passed as part of the options passed to the tiling method.

This is a breaking change with the following changes

1) The `scf::tileUsingSCFForOp` is renamed to `scf::tileUsingSCF`
2) The `scf::tileUsingSCFForallOp` is deprecated. The same
   functionality is obtained by using `scf::tileUsingSCF` and setting
   the loop type in `scf::SCFTilingOptions` passed into this method to
   `scf::SCFTilingOptions::LoopType::ForallOp` (using the
   `setLoopType` method).
3) The `scf::tileConsumerAndFusedProducerGreedilyUsingSCFForOp` is
   renamed to `scf::tileConsumerAndFuseProducerUsingSCF`. The use of
   the `controlFn` in `scf::SCFTileAndFuseOptions` allows implementing
   any strategy with the default callback implemeting the greedy fusion.
4) The `scf::SCFTilingResult` and `scf::SCFTileAndFuseResult` now use
   `SmallVector<LoopLikeOpInterface>`.
5) To make `scf::ForallOp` implement the parts of
   `LoopLikeOpInterface` needed, the `getOutputBlockArguments()`
   method is replaced with `getRegionIterArgs()`

These changes now bring the tiling and fusion capabilities using `scf.forall` on par with what was already supported by `scf.for`

MaheshRavishankar committed:
[mlir][TilingInterface] Use `LoopLikeOpInterface` in tiling using SCF to unify tiling with `scf.for` and `scf.forall`.

Using `LoopLikeOpInterface` as the basis for the implementation
unifies all the tiling logic for both `scf.for` and `scf.forall`. The
only difference is the actual loop generation.
Instead of many entry points for each loop type, the loop type is now
passed as part of the options passed to the tiling method.

This is a breaking change with the following changes

1) The `scf::tileUsingSCFForOp` is renamed to `scf::tileUsingSCF`
2) The `scf::tileUsingSCFForallOp` is deprecated. The same
   functionality is obtained by using `scf::tileUsingSCF` and setting
   the loop type in `scf::SCFTilingOptions` passed into this method to
   `scf::SCFTilingOptions::LoopType::ForallOp` (using the
   `setLoopType` method).
3) The `scf::tileConsumerAndFusedProducerGreedilyUsingSCFForOp` is
   renamed to `scf::tileConsumerAndFuseProducerUsingSCF`. The use of
   the `controlFn` in `scf::SCFTileAndFuseOptions` allows implementing
   any strategy with the default callback implemeting the greedy fusion.
4) The `scf::SCFTilingResult` and `scf::SCFTileAndFuseResult` now use
   `SmallVector<LoopLikeOpInterface>`.
5) To make `scf::ForallOp` implement the parts of
   `LoopLikeOpInterface` needed, the `getOutputBlockArguments()`
   method is replaced with `getRegionIterArgs()`

This change also introduces a new interface method for
`LoopLikeOpInterface`, that allows loop constructs to handle tiled
yields.

These changes now bring the tiling and fusion capabilities using
`scf.forall` on par with what was already supported

Abhishek-Varma reviewed (approved):
This is amazing @MaheshRavishankar !
Just one nit comment and CI related fix from my end.

matthias-springer commented on mlir/lib/Dialect/SCF/IR/SCF.cpp:
```diff
@@ -622,6 +626,47 @@ LogicalResult scf::ForallOp::promoteIfSingleIteration(RewriterBase &rewriter) {
+    OpBuilder::InsertionGuard g(rewriter);
+    auto terminator = cast<scf::InParallelOp>(newLoopBody->getTerminator());
+    rewriter.setInsertionPointToEnd(terminator.getBody());
+    newYieldValueFn(
```
I couldn't find the place in the code, but I think this lambda creates the `parallel_insert_slice` ops. It probably returns an empty `SmallVector`. This does not fit with the comment in `LoopLikeInterface.h`:

```c++
/// A function that returns the additional yielded values during
/// `replaceWithAdditionalYields`. `newBbArgs` are the newly added region
/// iter_args. This function should return as many values as there are block
/// arguments in `newBbArgs`.
using NewYieldValuesFn = std::function<SmallVector<Value>(
    OpBuilder &b, Location loc, ArrayRef<BlockArgument> newBbArgs)>;
```

It is a bit confusing that for `scf.for` we expect the function to return a list of `Value` and for `scf.forall` we expect the function to build new IR and not return anything.

How about we generalize `NewYieldValuesFn` to `RewriteTerminatorFn = std::function<void(RewriterBase &rewriter, Operation *terminator, ArrayRef<BlockArgument> newBbArgs)>`?

MaheshRavishankar replied:
Yeah, but the `scf.forall` does not return any values. I looked at place where `NewYieldValuesFn` is used and it is quite a bit. Want to limit the breaking changes here, but I can do a quick follow up on that one.

MaheshRavishankar replied:
Actually that might not work as well. Consider https://github.com/llvm/llvm-project/blob/5ca2d75f2046612978ba71c4b36714b2a0a01886/mlir/lib/Dialect/Linalg/Transforms/Hoisting.cpp#L193 . This is written in a loop agnostic way, but yield value function will have to handle `scf.forall` separately. I think there is an inherent mismatch between the behavior of `scf.forall` and other loop constructs in the sense it doesnt yield anything.

MaheshRavishankar replied:
Actually I do have an issue here (and I didnt see it cause I am also missing a test that will show up this issue). So good catch... Ill drop this part and do something different...

matthias-springer replied:
> I think there is an inherent mismatch between the behavior of `scf.forall` and other loop constructs in the sense it doesnt yield anything.

I ran into the same problem when cleaning up the `LoopLikeOpInterface` recently. I was able to add the interface to `scf.while`, but not `scf.forall` because of this issue.

I think we should generalize `NewYieldValuesFn` to `RewriteTerminatorFn`. That lambda can do whatever it wants.

```c++
/// A function that updates the terminator of a loop-like op during
/// `replaceWithAdditionalYields`. `newBbArgs` are the newly added region
/// iter_args. This function should rewrite the loop terminator (`terminator`)
/// such that the loop op verifies successfully.
using RewriteTerminatorFn =
    std::function<void(RewriterBase &rewriter, Operation *terminator, ArrayRef<BlockArgument> newBbArgs)>;
```

I think we also have to turn this interface function into an interface method:
```c++
    /// Append the specified additional "init" operands: replace this loop with
    /// a new loop that has the additional init operands. The loop body of this
    /// loop is moved over to the new loop.
    ///
    /// The newly added region iter_args are yielded from the loop.
    ::mlir::FailureOr<::mlir::LoopLikeOpInterface>
        replaceWithAdditionalIterOperands(::mlir::RewriterBase &rewriter,
                                          ::mlir::ValueRange newInitOperands,
                                          bool replaceInitOperandUsesInLoop)
```

For `scf.forall`, this interface could either:
- "return failure();"
- or: insert a parallel_insert_slice that inserts the newly added bbArg into itself (which is a no-op)
- or: remove this interface method entirely

matthias-springer replied:
The implementations of `yieldTiledValuesAndReplace` make sense to me. But I wouldn't put it in `LoopLikeOpInterface`.

`LoopLikeOpInterface` was meant to expose basic loop properties and low-level transformations of loop-like ops. The new `yieldTiledValuesAndReplace` interface method is tiling-specific and does not belong in `LoopLikeOpInterface` imo. How about putting `yieldTiledValuesAndReplace` in a new interface in `TilingInterface.td`? Then we have all tiling-related stuff in one place.

MaheshRavishankar replied:
I am not sure a whole interface for just one method is worth it. I think this method is optional for the most part. No op needs to implement it. It is needed only if it wants to work with tiling, which is nice cause all the loop construct has to do to work with tiling is implement this method. Id prefer to keep it in `LoopLikeOpInterface`. If you feel strongly about it, I can move it out of the interface.

matthias-springer reviewed (changes requested):
Looks good overall, just a small issue with `NewYieldValuesFn`.

chelini commented on mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:
```diff
@@ -135,93 +111,231 @@ static Operation *cloneOpAndUpdateDestinationArgs(RewriterBase &rewriter,
+  if (options.loopType == scf::SCFTilingOptions::LoopType::ForOp) {
+    return generateLoopNestUsingForOp(rewriter, loc, loopRanges, tileSizes,
+                                      destinationTensors, tiledBodyFn, loops);
+  } else if (options.loopType == scf::SCFTilingOptions::LoopType::ForallOp) {
```
We can use an if statement here as we have a return value. See: https://clang.llvm.org/extra/clang-tidy/checks/readability/else-after-return.html

chelini commented on mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:
```diff
@@ -230,48 +344,48 @@ static void addInitOperandsToLoopNest(
+  FailureOr<LoopLikeOpInterface> newInnerMostLoop =
+      innerMostLoop.replaceWithAdditionalYields(rewriter, newInitValues, false,
+                                                getNewYieldValsFn);
+  if (failed(newInnerMostLoop)) {
```
nit: drop braces here?

chelini commented on mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.td:
```diff
@@ -71,11 +71,33 @@ def TestTileUsingForallOp : Op<Transform_Dialect, "test.tile_using_forall",
   }];
 }
 
+def TestFuseUsingForallOp : Op<Transform_Dialect, "test.fuse_using_forall",
```
maybe: `fuse_using_forall_greedily` ?

MaheshRavishankar replied:
just using this as a mirror for `fuse` op for testing. I want to fold this op into a proper transform op, I just dont know the final form of the operation. So maybe ill leave it as is for now?

chelini commented on mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.cpp:
```diff
@@ -235,6 +236,97 @@ void transform::TestTileUsingForallOp::getEffects(
+
+DiagnosedSilenceableFailure transform::TestFuseUsingForallOp::apply(
+    transform::TransformRewriter &rewriter,
+    mlir::transform::TransformResults &transformResults,
```
let's drop mlir:: please.

chelini commented on mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.cpp:
```diff
@@ -235,6 +236,97 @@ void transform::TestTileUsingForallOp::getEffects(
+      for (OpResult res : toReplace->getResults())
+        if (auto replacement = tiledResults->replacements.lookup(res))
+          rewriter.replaceAllUsesWith(res, replacement);
+      if (toReplace->use_empty()) {
```
nit: drop braces

chelini commented on mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:
```diff
@@ -464,50 +564,73 @@ mlir::scf::tileReductionUsingScf(RewriterBase &b,
+  FailureOr<TilingResult> tilingResult =
+      generateLoopNest(b, loc, options, iterationDomain, tileSizesVector,
+                       destinationTensors, innerTileLoopBodyFn, loops);
+  if (failed(tilingResult)) {
```
drop braces?

chelini commented on mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:
```diff
@@ -464,50 +564,73 @@ mlir::scf::tileReductionUsingScf(RewriterBase &b,
+    {
+      int materializedLoopNum = 0;
+      for (auto [tileSize, loopRange] :
+           llvm::zip(tileSizesVector, iterationDomain)) {
```
zip_equal?

chelini commented on mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:
```diff
@@ -464,50 +564,73 @@ mlir::scf::tileReductionUsingScf(RewriterBase &b,
+    b.eraseOp(clonedOp);
+
+    // 4d. Compute the offsets and sizes needed to insert the result of the
+    // tiled
```
wired comment layout.

chelini commented on mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:
```diff
@@ -288,145 +402,131 @@ mlir::scf::tileUsingSCFForOp(RewriterBase &rewriter, TilingInterface op,
-      cloneOpAndUpdateDestinationArgs(rewriter, op, clonedOpDestination));
+    // 5b. Early return cloned op if tiling is not happening. We can not return
+    // the original op because it could lead to
+    // `rewriter.replaceOp(op, op->getResults())` and user would get crash.
```
nit: users

chelini replied:
Where this `replaceOp` happening? I would expect we could avoid cloning the op if the tile sizes are all zeros.

chelini commented on mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:
```diff
@@ -288,145 +402,131 @@ mlir::scf::tileUsingSCFForOp(RewriterBase &rewriter, TilingInterface op,
+    // 5b. Early return cloned op if tiling is not happening. We can not return
+    // the original op because it could lead to
+    // `rewriter.replaceOp(op, op->getResults())` and user would get crash.
+    if (llvm::all_of(tileSizes, isZeroIndex)) {
```
nit: drop braces.

chelini commented on mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:
```diff
@@ -288,145 +402,131 @@ mlir::scf::tileUsingSCFForOp(RewriterBase &rewriter, TilingInterface op,
+    SmallVector<OpFoldResult> offsets, sizes;
+    {
+      int materializedLoopNum = 0;
+      for (auto [tileSize, loopRange] : llvm::zip(tileSizes, iterationDomain)) {
```
zip_equal?

chelini commented on mlir/lib/Dialect/SCF/IR/SCF.cpp:
```diff
@@ -1630,9 +1675,8 @@ struct FoldTensorCastOfOutputIntoForallOp
-    for (auto [yieldingOp, outputBlockArg] :
-         llvm::zip(terminator.getYieldingOps(),
-                   newForallOp.getOutputBlockArguments())) {
+    for (auto [yieldingOp, outputBlockArg] : llvm::zip(
```
zip_equal.

chelini replied:
Can we use zip_equal, here?

chelini commented on mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:
```diff
@@ -135,93 +111,231 @@ static Operation *cloneOpAndUpdateDestinationArgs(RewriterBase &rewriter,
+    return rewriter.notifyMatchFailure(
+        loc, "failed to generate inner tile loop body");
+  }
+  if (loops.empty()) {
```
nit braces

chelini commented on mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:
```diff
@@ -135,93 +111,231 @@ static Operation *cloneOpAndUpdateDestinationArgs(RewriterBase &rewriter,
+  SmallVector<OpFoldResult> offsets(loopRanges.size()),
+      sizes(loopRanges.size());
+
+  for (auto [tileSize, loopRange] : llvm::zip(tileSizes, loopRanges)) {
```
zip_equal

chelini commented on mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:
```diff
@@ -135,93 +111,231 @@ static Operation *cloneOpAndUpdateDestinationArgs(RewriterBase &rewriter,
+  assert(!lbs.empty() && "Expected at least one loop range");
+
+  std::optional<ArrayAttr> mappingAttr;
+  if (!mappingVector.empty()) {
```
braces here and a bit below (line 261).

chelini reviewed (commented):
Thanks for doing this +1

nicolasvasilache commented on mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:
```diff
@@ -135,93 +111,231 @@ static Operation *cloneOpAndUpdateDestinationArgs(RewriterBase &rewriter,
+///    populated.
+/// The `TilingResult` returned by calling `tileLoopBodyFn` is returned back
+/// to the caller.
+static FailureOr<TilingResult> generateLoopNestUsingForOp(
```
How does this differ / resemble older code that lowers Linalg to loops?

MaheshRavishankar replied:
you mean lowering to loops code? Maybe.. Ill look into that.

nicolasvasilache reviewed (dismissed):
This is the followup from #72178, right?
Let's mention it in the commit message so we can refer back to it later, had to do a bit of spelunking to find it.

Overall comment:

So IIRC, #72178 adopted the logic of tiling `scf.forall` from `LinalgTransformOps.cpp` and/or `Linalg/Transforms/Tiling.cpp` to redo `scf.for`.

This was a welcome refactoring and this PR is a nice continuation, nice!

To proceed confidently, I would now like to see red diffs in `LinalgTransformOps.cpp` and `Linalg/Transforms/Tiling.cpp` to see older implementations be actually replaced: atm this seems it is still adding more code in parallel to the old implementation.

For instance, I still see LinalgTransformOps.cpp calling code from `Linalg/Transforms/Tiling.cpp`:
```
  FailureOr<linalg::ForallTilingResult> maybeTilingResult = failure();
  if (!mixedNumThreads.empty()) {
    maybeTilingResult =
        linalg::tileToForallOp(rewriter, tileableOp, mixedNumThreads, mapping);
  } else {
    maybeTilingResult = linalg::tileToForallOpUsingTileSizes(
        rewriter, tileableOp, mixedTileSizes, mapping);
  }
```

Do you have a followup PR which cleans up the old?

Note, as part of this, I would also like to see something concrete on the `num_threads` discussion.

Besides the meta-comment that we now need to resolve, I'll be able to review this early next week.

Thanks!

MaheshRavishankar commented:
> This is the followup from #72178, right? Let's mention it in the commit message so we can refer back to it later, had to do a bit of spelunking to find it.

It is a change from a long list of related changes, but yes, https://github.com/llvm/llvm-project/pull/72178 is probably the start of it. Updated the comment anyway.

> 
> Overall comment:
> 
> So IIRC, #72178 adopted the logic of tiling `scf.forall` from `LinalgTransformOps.cpp` and/or `Linalg/Transforms/Tiling.cpp` to redo `scf.for`.
> 
> This was a welcome refactoring and this PR is a nice continuation, nice!
> 
> To proceed confidently, I would now like to see red diffs in `LinalgTransformOps.cpp` and `Linalg/Transforms/Tiling.cpp` to see older implementations be actually replaced: atm this seems it is still adding more code in parallel to the old implementation.

Yes, that is the goal, but I want to proceed incrementally. With this change, the same tests that work with `scf.for` work with `scf.forall` (so interchange, scalars, etc. etc) and avoids duplicating code for tiling using the two loop constructs. The next step for me is to move over the tiling logic that exists currently in `Linalg/Tiling.cpp` and `LinalgTransformOps.cpp` to this and fix as needed.

There is lot of context here spread all over the place w.r.t tiling. I want to unify all of it, but I dont think I can do that in one step. I would rather incrementally build up the required functionality in one place and deprecate/move pieces as needed. I do plan to do this in short order though..

> 
> For instance, I still see LinalgTransformOps.cpp calling code from `Linalg/Transforms/Tiling.cpp`:
> 
> ```
>   FailureOr<linalg::ForallTilingResult> maybeTilingResult = failure();
>   if (!mixedNumThreads.empty()) {
>     maybeTilingResult =
>         linalg::tileToForallOp(rewriter, tileableOp, mixedNumThreads, mapping);
>   } else {
>     maybeTilingResult = linalg::tileToForallOpUsingTileSizes(
>         rewriter, tileableOp, mixedTileSizes, mapping);
>   }
> ```
> 
> Do you have a followup PR which cleans up the old?
> 
> Note, as part of this, I would also like to see something concrete on the `num_threads` discussion.

`num_threads` is the missing piece for me. I had asked this previously as to what is the difference between using loops of the form `[0, num_threads)` with the induction variable being some thread ID, vs. induction variable being `[lb, ub)` with `step`. It seems to be that the use of thread ID needs to come in only at the point of "resolution" of `scf.forall` but I maybe missing something. Again, treating it incrementally. Now with this PR I have the basics to start unwinding some of this.

> 
> Besides the meta-comment that we now need to resolve, I'll be able to review this early next week.
> 
> Thanks!

Thanks!

MaheshRavishankar commented:
@matthias-springer I am having the following three tests failing for me, for reasons that seem unrelated to this PR (related by the fact that this is happening with this PR, but unrelated in the sense that its the pad hoisting code failing even though the tiling is the same).
```
[build] Failed Tests (3):
[build]   MLIR :: Dialect/Linalg/transform-op-hoist-pad-build-packing-loop-nest.mlir
[build]   MLIR :: Dialect/Linalg/transform-op-hoist-pad.mlir
[build]   MLIR :: Dialect/Linalg/transform-op-peel-and-vectorize.mlir
```

I tried to triage this, but the issue seems deep in the weeds. I will continue looking, but if you have any pointers, that would be very helpful.

MaheshRavishankar commented:
> @matthias-springer I am having the following three tests failing for me, for reasons that seem unrelated to this PR (related by the fact that this is happening with this PR, but unrelated in the sense that its the pad hoisting code failing even though the tiling is the same).
> 
> ```
> [build] Failed Tests (3):
> [build]   MLIR :: Dialect/Linalg/transform-op-hoist-pad-build-packing-loop-nest.mlir
> [build]   MLIR :: Dialect/Linalg/transform-op-hoist-pad.mlir
> [build]   MLIR :: Dialect/Linalg/transform-op-peel-and-vectorize.mlir
> ```
> 
> I tried to triage this, but the issue seems deep in the weeds. I will continue looking, but if you have any pointers, that would be very helpful.

Never mind. Figured it out.. the hoisting logic needed some LICM to kick to work. Fixed now.

matthias-springer commented on mlir/lib/Dialect/SCF/IR/SCF.cpp:
```diff
@@ -622,6 +626,47 @@ LogicalResult scf::ForallOp::promoteIfSingleIteration(RewriterBase &rewriter) {
+  return getBody()->getArguments().drop_front(getRank());
+}
+
+MutableArrayRef<OpOperand> ForallOp::getInitsMutable() {
```
I think you are missing a `getYieldedValuesMutable` implementation for `ForallOp`.

This is what the interface documentation says:
```
    Loop-carried variables can be exposed through this interface. There are
    3 components to a loop-carried variable.
    - The "region iter_arg" is the block argument of the entry block that
      represents the loop-carried variable in each iteration.
    - The "init value" is an operand of the loop op that serves as the initial
      region iter_arg value for the first iteration (if any).
    - The "yielded" value is the value that is forwarded from one iteration to
      serve as the region iter_arg of the next iteration.

    If one of the respective interface methods is implemented, so must the other
    two. The interface verifier ensures that the number of types of the region
    iter_args, init values and yielded values match.
```

This is checked by the op verifier, so you should be seeing verification failures.

We could have `getYieldedValuesMutable` return a `std::optional<::llvm::MutableArrayRef<::mlir::OpOperand>>` to account for the fact that some loops do not have yielding semantics. (Same as we do for `getLoopResults`.)

MaheshRavishankar replied:
> I think you are missing a `getYieldedValuesMutable` implementation for `ForallOp`.
> 
> This is what the interface documentation says:
> 
> ```
>     Loop-carried variables can be exposed through this interface. There are
>     3 components to a loop-carried variable.
>     - The "region iter_arg" is the block argument of the entry block that
>       represents the loop-carried variable in each iteration.
>     - The "init value" is an operand of the loop op that serves as the initial
>       region iter_arg value for the first iteration (if any).
>     - The "yielded" value is the value that is forwarded from one iteration to
>       serve as the region iter_arg of the next iteration.
> 
>     If one of the respective interface methods is implemented, so must the other
>     two. The interface verifier ensures that the number of types of the region
>     iter_args, init values and yielded values match.
> ```

Obviously that definition does not work for `scf.forall`. Might need to update this since the `scf.forall` does not yield any value. The region iter arg and init value can stay though.

> 
> This is checked by the op verifier, so you should be seeing verification failures.
> 

I did. I fixed the verifier also, but I think as you note below, `getYieldedValuesMutable` should return an `std::optional<..>`


> On second thought, scf.forall should not implement getRegionIterArgs at all. The region iter_args in the loop like op interface are loop-carried variables, but the scf.forall does not have any loop-carried variables.

We probably dont need that.. the region iter args can still be tied to init. If there is no yield value, then the verifier can handle it appropriately.

joker-eph commented on mlir/include/mlir/Interfaces/LoopLikeInterface.h:
```diff
@@ -25,6 +25,28 @@ class RewriterBase;
+/// - `resultStrides` is of the same size as `tiledValues` and represents
+///   the strides to use when inserting corresponding element from `tiledValues`
+///   into the element from `newBbArgs`.
+using YieldTiledValuesFn = std::function<LogicalResult(
```
It isn't stored, should be a function_ref.

MaheshRavishankar replied:
I can change it, but the `NewYieldsValueFn` is also a `std::function`, so was being consistent with that.

joker-eph replied:
Fix both?

MaheshRavishankar replied:
Ok, trying, but I have hit some really bad footguns with `function_ref` in the past. I'd rather just pass everything by value.

joker-eph commented on mlir/include/mlir/Interfaces/LoopLikeInterface.td:
```diff
@@ -218,6 +218,19 @@ def LoopLikeOpInterface : OpInterface<"LoopLikeOpInterface"> {
     >,
+    InterfaceMethod<[{
+        TODO
+      }],
```
Doc?

MaheshRavishankar replied:
Oops! Sorry forgot. Added documentation.

nicolasvasilache replied:
great, thanks for factoring out this functionality.

joker-eph commented on mlir/lib/Dialect/SCF/IR/SCF.cpp:
```diff
@@ -584,6 +588,63 @@ ForOp::replaceWithAdditionalYields(RewriterBase &rewriter,
+
+  auto inits = llvm::to_vector(getInitArgs());
+  inits.append(newInitOperands.begin(), newInitOperands.end());
+  scf::ForOp newLoop = rewriter.create<scf::ForOp>(
```
No need for scf::  here: we're inside the ForOp itself!

joker-eph reviewed (commented):
I agree with @matthias-springer : it's not clear to me how this belongs to the LoopLikeInterface!
(but also it's not documented, so I don't even know what this is doing really)

MaheshRavishankar committed:
Make `getYieldedValuesMutable` return an `std::optional<llvm::MutableArrayRef>`.

MaheshRavishankar committed:
Add description for `yieldTiledValuesAndReplace`.

MaheshRavishankar committed:
Attempt to fix Windows build error.

MaheshRavishankar committed:
Address comments (round 1).

joker-eph commented on mlir/include/mlir/Interfaces/LoopLikeInterface.td:
```diff
@@ -218,6 +221,53 @@ def LoopLikeOpInterface : OpInterface<"LoopLikeOpInterface"> {
+
+        This method is similar to `replaceWithAdditionalYields` but instead of
+        returning the value that is actually yielded, this returns the tiles of
+        the values that are yielded. This allows for unified handling of opreations
```
```suggestion
        the values that are yielded. This allows for unified handling of operations
```

joker-eph commented on mlir/include/mlir/Interfaces/LoopLikeInterface.td:
```diff
@@ -218,6 +221,53 @@ def LoopLikeOpInterface : OpInterface<"LoopLikeOpInterface"> {
+        This method is similar to `replaceWithAdditionalYields` but instead of
+        returning the value that is actually yielded, this returns the tiles of
+        the values that are yielded. This allows for unified handling of opreations
+        like `scf.forall` which dont yield a value from the loop, but instead
```
```suggestion
        like `scf.forall` which don't yield a value from the loop, but instead
```

joker-eph commented on mlir/include/mlir/Interfaces/LoopLikeInterface.td:
````diff
@@ -218,6 +221,53 @@ def LoopLikeOpInterface : OpInterface<"LoopLikeOpInterface"> {
+        ```mlir
+        %0 = scf.forall ... shared_outs(%arg0 = %arg1) {
+          ...
+          %tiled_value
````
```suggestion
          %tiled_value = ...
```

joker-eph commented on mlir/include/mlir/Interfaces/LoopLikeInterface.td:
````diff
@@ -218,6 +221,53 @@ def LoopLikeOpInterface : OpInterface<"LoopLikeOpInterface"> {
+        }
+        ```
+
+        So for the caller, the tiled value (`%tiled_values`) and the offsets
````
```suggestion
        So for the caller, the tiled value (`%tiled_value`) and the offsets
```

MaheshRavishankar commented:
> I agree with @matthias-springer : it's not clear to me how this belongs to the LoopLikeInterface! (but also it's not documented, so I don't even know what this is doing really)

I think it depends on what `LoopLikeInterface` should cover. Should `scf.forall` be part of the interface or not, cause it doesnt actually yield any value. This loops construct allows for parallel insertion of tiles into the result. So if something that has to unify how yielded values are handled, this is one way I can think of. Of course this goes hand-in-hand with tiling cause that is one way to create parallel work. See example in the documentation I added now (sorry I missed it earlier).
Also if a loop doesnt know how to handle tiles, it can always return a failure. So it isnt a required method to implement. So I think it makes sense to add it to the interface. This is similar to the `replaceWithAdditionalYields` method. This method does not make sense for `scf.forall` so it should return a failure for that method.

joker-eph commented on mlir/include/mlir/Interfaces/LoopLikeInterface.td:
```diff
@@ -218,6 +221,53 @@ def LoopLikeOpInterface : OpInterface<"LoopLikeOpInterface"> {
+        this loop is moved over to the new loop.
+
+        This method is similar to `replaceWithAdditionalYields` but instead of
+        returning the value that is actually yielded, this returns the tiles of
```
This does not seem to correctly describe the method: it returns a LoopLikeOpInterface but here you're saying "this returns the tiles of the values that are yielded"?

MaheshRavishankar committed:
Fix flang errors due to change in `getYieldedValuesMutable` signature.

MaheshRavishankar committed:
Fix code-formatting error.

MaheshRavishankar committed:
Fix documentation errors.

matthias-springer commented on mlir/lib/Interfaces/LoopLikeInterface.cpp:
```diff
@@ -63,8 +63,9 @@ LogicalResult detail::verifyLoopLikeOpInterface(Operation *op) {
            << " != " << loopLikeOp.getRegionIterArgs().size();
-  if (loopLikeOp.getRegionIterArgs().size() !=
-      loopLikeOp.getYieldedValues().size())
+  if (!loopLikeOp.getYieldedValues().empty() &&
```
Actually, there is a better way to check without `std::nullopt`:
```
if (!loopLikeOp.getYieldedValuesMutable().getOwner() && ...)
```

MaheshRavishankar committed:
Use std::function_ref.

nicolasvasilache commented on mlir/include/mlir/Dialect/SCF/IR/SCFOps.td:
```diff
@@ -135,10 +135,10 @@ def ExecuteRegionOp : SCF_Op<"execute_region", [
-        "getSingleStep", "getSingleUpperBound", "getYieldedValuesMutable",
-        "getLoopResults", "promoteIfSingleIteration",
-        "replaceWithAdditionalYields"]>,
+       ["getInitsMutable", "getRegionIterArgs", "getSingleInductionVar", 
```
nit: can we make the list 1 per line to improve readability ?

Also, `getLoopResults` is not properly ordered.

nicolasvasilache commented on mlir/include/mlir/Dialect/SCF/IR/SCFOps.td:
```diff
@@ -304,8 +300,9 @@ def ForallOp : SCF_Op<"forall", [
        DeclareOpInterfaceMethods<LoopLikeOpInterface,
-          ["promoteIfSingleIteration", "getSingleInductionVar",
-          "getSingleLowerBound", "getSingleUpperBound", "getSingleStep"]>,
+          ["getInitsMutable", "getRegionIterArgs", "getSingleInductionVar", 
```
nit: same as above

nicolasvasilache commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -485,8 +485,8 @@ transform::FuseOp::apply(transform::TransformRewriter &rewriter,
           -> FailureOr<scf::SCFTileAndFuseResult> {
-        return tileConsumerAndFuseProducerGreedilyUsingSCFForOp(
-            rewriter, tilingInterfaceOp, tileAndFuseOptions);
+        return tileConsumerAndFuseProducersUsingSCF(rewriter, tilingInterfaceOp,
```
nit: is the 'UsingSCF' part of the name still load-bearing now that we don't distinguish between `scf.for` and `scf.forall` anymore ?

MaheshRavishankar replied:
Well, good question. This is in the SCF dialect... So I wanted to leave the "SCF" part. I dont have a strong preference, but this is living in SCF dialect.

matthias-springer commented:
Just to clarify, I think this is a good refactoring, I'd just like to see it in the same place as the remaining tiling-related code.

> I think it depends on what `LoopLikeInterface` should cover.

Tiling is a quite complex transformation. Implementations of `YieldTiledValuesFn` have to create `tensor.insert_slice` / `tensor.parallel_insert_slice` ops. And tiling only makes sense for specific types (TensorType). At the moment, tiling is nicely isolated in `TilingInterface.td`.

nicolasvasilache commented on mlir/lib/Dialect/SCF/IR/SCF.cpp:
```diff
@@ -584,6 +588,63 @@ ForOp::replaceWithAdditionalYields(RewriterBase &rewriter,
+  if (failed(yieldTiledValuesFn(rewriter, getLoc(), newLoop.getInductionVar(),
+                                newRegionIterArgs, tiledValues, resultOffsets,
+                                resultSizes))) {
+    return rewriter.notifyMatchFailure(getOperation(),
```
Careful here as we have already created IR, so depending how this is used (today and in the future) we may or may not violate invariants (e.g. rewriter returns failure but IR has been modified).

Not 100% sure this is/will be a problem but:
1. Consider cleaning up the newLoop here ?
2. Also consider adding documentation to the lambda (e.g. either it must succeed or clean up after itself)?

MaheshRavishankar replied:
Done.

nicolasvasilache commented on mlir/lib/Dialect/SCF/IR/SCF.cpp:
```diff
@@ -584,6 +588,63 @@ ForOp::replaceWithAdditionalYields(RewriterBase &rewriter,
+                                       "failed to get tiled values");
+  }
+
+  if (tiledValues.size() != resultOffsets.size() ||
```
In case my comment just above is relevant, this has the same issue.
Consider making the writer of `yieldTiledValuesFn` guarantee this invariant and use an assert here instead?

MaheshRavishankar replied:
Done

joker-eph commented on mlir/lib/Dialect/SCF/IR/SCF.cpp:
```diff
@@ -584,6 +588,63 @@ ForOp::replaceWithAdditionalYields(RewriterBase &rewriter,
 
+FailureOr<LoopLikeOpInterface> ForOp::yieldTiledValuesAndReplace(
+    RewriterBase &rewriter, ValueRange newInitOperands,
+    const YieldTiledValuesFn &yieldTiledValuesFn) {
```
```suggestion
    YieldTiledValuesFn yieldTiledValuesFn) {
```

MaheshRavishankar committed:
Address comments (round 2)

- Remove passing as `const &` for the `llvm::function_ref`.
- Fix ordering of interface methods.

MaheshRavishankar committed:
Drop `yieldTiledValuesAndReplace` as an interface method.

MaheshRavishankar committed:
Try not using `llvm::function_ref`.

MaheshRavishankar committed:
Address remaining comments (round 3)

matthias-springer commented on mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h:
```diff
@@ -71,23 +80,17 @@ struct SCFTilingResult {
 };
 
 /// Method to tile an op that implements the `TilingInterface` using
 /// `scf.for` for iterating over the tiles.
```
Comment is outdated: This function may geneate `scf.for` or `scf.forall`.

MaheshRavishankar commented:
@nicolasvasilache could you take a look again?

nicolasvasilache commented:
This PR looks fine to me, thanks!

I'd like to see the cleanup work truly start before we land more code.

```
There is lot of context here spread all over the place w.r.t tiling. I want to unify all of it, but I dont think I can do that in one step. I would rather incrementally build up the required functionality in one place and deprecate/move pieces as needed. I do plan to do this in short order though..
```

To be clear, I did not ask for a making this PR more complex but specifically asked for a followup PR to see that work has truly unambiguously started on the deprecation of the legacy implementation (I already voiced this back in Nov: https://github.com/llvm/llvm-project/pull/72178#issuecomment-1810872756).

Now is the time to see concrete movement on the deprecation before we land more code.

MaheshRavishankar commented:
> This PR looks fine to me, thanks!
> 
> I'd like to see the cleanup work truly start before we land more code.
> 
> ```
> There is lot of context here spread all over the place w.r.t tiling. I want to unify all of it, but I dont think I can do that in one step. I would rather incrementally build up the required functionality in one place and deprecate/move pieces as needed. I do plan to do this in short order though..
> ```
> 
> To be clear, I did not ask for a making this PR more complex but specifically asked for a followup PR to see that work has truly unambiguously started on the deprecation of the legacy implementation (I already voiced this back in Nov: [#72178 (comment)](https://github.com/llvm/llvm-project/pull/72178#issuecomment-1810872756)).
> 
> Now is the time to see concrete movement on the deprecation before we land more code.

I agree, my next step is to move the transform dialect op to use this implementation. So thats top on the list.

Commit:
Fix build failure hit downstream.

Commit:
Fix bazel build.
